bes_publisher/buildevent/service.go

```diff
@@ -0,0 +1,154 @@
+		Namespace: "bes_publisher",
+		Name:      "bes_event_count",
+		Help:      "Number of BES events, grouped by how they were handled",
+	},
+		[]string{
+			"event_type",
+			"outcome",
+		},
+	)
+)
+
+// oneofType returns a friendly string for a protobuf oneof type to use in
+// logging/metrics.
+//
+// Converts a string like `*proto.BuildEvent_BuildMetadata` to `BuildMetadata`.
+func oneofType(msg any) string {
+	ret := "<unknown>"
+	elems := strings.Split(fmt.Sprintf("%T", msg), "_")
+	if len(elems) < 2 {
+		return ret
+	}
+	return elems[1]
+}
+
+// bazelEventFrom returns a bazel BES BuildEvent from the given event stream
+// message, or an error if the message is of a different type.
+func bazelEventFrom(req *bpb.PublishBuildToolEventStreamRequest) (*bes.BuildEvent, error) {
+	switch event := req.GetOrderedBuildEvent().GetEvent().Event.(type) {
+	default:
+		metricBuildEventProtocolEventCount.WithLabelValues(oneofType(event), "dropped").Inc()
+		return nil, fmt.Errorf("not handling unknown BEP event type: %T", event)
+	case *bpb.BuildEvent_BazelEvent:
+		buildEvent := &bes.BuildEvent{}
+		if err := ptypes.UnmarshalAny(event.BazelEvent, buildEvent); err != nil {
+			metricBuildEventProtocolEventCount.WithLabelValues(oneofType(event), "parse_fail").Inc()
+			return nil, fmt.Errorf("failed to unmarshal embedded BazelEvent: %w", err)
+		}
+		metricBuildEventProtocolEventCount.WithLabelValues(oneofType(event), "ok").Inc()
+		return buildEvent, nil
+	}
+}
+
+// Service implements the Build Event Protocol service.
+type Service struct {
+}
+
+// PublishLifecycleEvent records the BEP lifecycle events seen in a metric, and
+// not much else.
+func (s *Service) PublishLifecycleEvent(ctx context.Context, req *bpb.PublishLifecycleEventRequest) (*emptypb.Empty, error) {
+	glog.V(2).Infof("# BEP LifecycleEvent message:\n%s", prototext.Format(req))
+	metricLifecycleEventCount.WithLabelValues(oneofType(req.GetBuildEvent().GetEvent().GetEvent()), "dropped").Inc()
+	return &emptypb.Empty{}, nil
+}
+
+// PublishBuildToolEventStream handles all the Bazel BES messages seen.
+func (s *Service) PublishBuildToolEventStream(stream bpb.PublishBuildEvent_PublishBuildToolEventStreamServer) (retErr error) {
+	bs := &buildStream{
+		stream: stream,
+	}
+	return bs.handleMessages()
+}
+
+// buildStream wraps a single stream (for a single build) so that it can
+// aggregate state seen across the entire stream, such as invocation ID and
+// build type.
+type buildStream struct {
+	stream bpb.PublishBuildEvent_PublishBuildToolEventStreamServer
+}
+
+// handleMessages handles all the messages on the stream, and then returns an
+// error if it encounters a non-EOF error while exhausting the stream.
+func (b *buildStream) handleMessages() error {
+	for {
+		req, err := b.stream.Recv()
+		if errors.Is(err, io.EOF) {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+		if err := b.handleEvent(req); err != nil {
```

Comment:
Dropping the error here. maybe log it or inc a metric?

Reply:
`handleEvent` itself has metrics, and `bazelEventFrom` inside that also has a metric recording their outcomes